File: app/src/androidTest/java/org/tasks/jobs/BackupServiceTests.java
/*
 * Copyright (c) 2012 Todoroo Inc
 *
 * See the file "LICENSE" for the full license governing this code.
 */

package org.tasks.jobs;

import static androidx.test.InstrumentationRegistry.getTargetContext;
import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertTrue;

import android.net.Uri;
import androidx.test.runner.AndroidJUnit4;
import com.todoroo.astrid.dao.TaskDao;
import com.todoroo.astrid.data.Task;
import java.io.File;
import java.io.IOException;
import javax.inject.Inject;
import org.junit.After;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.tasks.R;
import org.tasks.backup.TasksJsonExporter;
import org.tasks.backup.TasksJsonExporter.ExportType;
import org.tasks.injection.InjectingTestCase;
import org.tasks.injection.TestComponent;
import org.tasks.preferences.Preferences;

@RunWith(AndroidJUnit4.class)
public class BackupServiceTests extends InjectingTestCase {

  @Inject TasksJsonExporter jsonExporter;
  @Inject TaskDao taskDao;
  @Inject Preferences preferences;
  private File temporaryDirectory = null;

  @Override
  public void setUp() {
    super.setUp();

    try {
      temporaryDirectory = File.createTempFile("backup", Long.toString(System.nanoTime()));
    } catch (IOException e) {
      throw new RuntimeException(e);
    }

    if (!(temporaryDirectory.delete())) {
      throw new RuntimeException(
          "Could not delete temp file: " + temporaryDirectory.getAbsolutePath());
    }
    if (!(temporaryDirectory.mkdir())) {
      throw new RuntimeException(
          "Could not create temp directory: " + temporaryDirectory.getAbsolutePath());
    }

    preferences.setUri(R.string.p_backup_dir, Uri.fromFile(temporaryDirectory));

    // make a temporary task
    Task task = new Task();
    task.setTitle("helicopter");
    taskDao.createNew(task);
  }

  @Override
  protected void inject(TestComponent component) {
    component.inject(this);
  }

  @After
  public void tearDown() {
    if (temporaryDirectory != null) {
      for (File file : temporaryDirectory.listFiles()) {
        file.delete();
      }
      temporaryDirectory.delete();
    }
  }

  @Test
  public void testBackup() {
    assertEquals(0, temporaryDirectory.list().length);

    jsonExporter.exportTasks(getTargetContext(), ExportType.EXPORT_TYPE_SERVICE, null);

    // assert file created
    File[] files = temporaryDirectory.listFiles();
    assertEquals(1, files.length);
    assertTrue(files[0].getName().matches(BackupWork.BACKUP_FILE_NAME_REGEX));
  }
}

File: app/src/androidTest/java/org/tasks/jobs/BackupServiceTests.kt
/*
 * Copyright (c) 2012 Todoroo Inc
 *
 * See the file "LICENSE" for the full license governing this code.
 */
package org.tasks.jobs

import android.net.Uri
import androidx.test.InstrumentationRegistry
import com.todoroo.astrid.dao.TaskDao
import org.tasks.data.entity.Task
import dagger.hilt.android.testing.HiltAndroidTest
import dagger.hilt.android.testing.UninstallModules
import kotlinx.coroutines.runBlocking
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.tasks.R
import org.tasks.backup.BackupConstants.BACKUP_CLEANUP_MATCHER
import org.tasks.backup.TasksJsonExporter
import org.tasks.backup.TasksJsonExporter.ExportType
import org.tasks.injection.InjectingTestCase
import org.tasks.injection.ProductionModule
import org.tasks.preferences.Preferences
import java.io.File
import java.io.IOException
import javax.inject.Inject

@UninstallModules(ProductionModule::class)
@HiltAndroidTest
class BackupServiceTests : InjectingTestCase() {
    @Inject lateinit var jsonExporter: TasksJsonExporter
    @Inject lateinit var taskDao: TaskDao
    @Inject lateinit var preferences: Preferences
    private lateinit var temporaryDirectory: File

    @Before
    override fun setUp() {
        runBlocking {
            super.setUp()
            temporaryDirectory = try {
                File.createTempFile("backup", System.nanoTime().toString())
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
            if (!temporaryDirectory.delete()) {
                throw RuntimeException(
                        "Could not delete temp file: " + temporaryDirectory.absolutePath)
            }
            if (!temporaryDirectory.mkdir()) {
                throw RuntimeException(
                        "Could not create temp directory: " + temporaryDirectory.absolutePath)
            }
            preferences.setUri(R.string.p_backup_dir, Uri.fromFile(temporaryDirectory))

            // make a temporary task
            val task = Task()
            task.title = "helicopter"
            taskDao.createNew(task)
        }
    }

    @After
    fun tearDown() {
        for (file in temporaryDirectory.listFiles()!!) {
            file.delete()
        }
        temporaryDirectory.delete()
    }

    @Test
    fun testBackup() = runBlocking {
        assertEquals(0, temporaryDirectory.list()!!.size)
        jsonExporter.exportTasks(InstrumentationRegistry.getTargetContext(), ExportType.EXPORT_TYPE_SERVICE, null)

        // assert file created
        val files = temporaryDirectory.listFiles()
        assertEquals(1, files!!.size)
        assertTrue(files[0].name.matches(BACKUP_CLEANUP_MATCHER))
    }
}